DOC: update the pandas.errors.DtypeWarning docstring

hissashirocha opened the pull request:
Checklist for the pandas documentation sprint (ignore this if you are doing
an unrelated PR):

- [X] PR title is "DOC: update the <your-function-or-method> docstring"
- [X] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [X] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [X] It has been proofread on language by another sprint participant

Please include the output of the validation script below between the "```" ticks:

```
################################################################################
#################### Docstring (pandas.errors.DtypeWarning) ####################
################################################################################

Warning raised when reading different dtypes in a column from a file.

Raised for a dtype incompatibility. This can happen whenever `read_csv`
or `read_table` encounter non-uniform dtypes in a column(s) of a given
CSV file.

See Also
--------
pandas.read_csv : Read CSV (comma-separated) file into a DataFrame.
pandas.read_table : Read general delimited file into a DataFrame.

Notes
-----
This warning is issued when dealing with larger files because the dtype
checking happens per chunk read.

Despite the warning, the CSV file is read with mixed types in a single
column which will be an object type. See the examples below to better
understand this issue.

Examples
--------
This example creates and reads a large CSV file with a column that contains
`int` and `str`.

>>> df = pd.DataFrame({'a':['1']*100000 + ['X']*100000 + ['1']*100000,
...                    'b':['b']*300000})
>>> df.to_csv('test.csv', index=False)
>>> df2 = pd.read_csv('test.csv')
>>> DtypeWarning: Columns (0) have mixed types... # doctest: +SKIP

Important to notice that df2 will contain both `str` and `int` for the
same input, '1'.

>>> df2.iloc[262140, 0]
'1'
>>> type(df2.iloc[262140, 0])
<class 'str'>
>>> df2.iloc[262150, 0]
1
>>> type(df2.iloc[262150, 0])
<class 'int'>

One way to solve this issue is using the parameter `converters` in the
`read_csv` and `read_table` functions to explicit the conversion:

>>> df2 = pd.read_csv('test', sep=' ', converters={'a': str})
scripts/validate_docstrings.py:268: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  runner.run(test)

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	No returns section found
```

hissashirocha committed:
Improving docstring for DtypeWarning.

hissashirocha committed:
Improving docstring for DtypeWarning.

hissashirocha committed:
Improving docstring for DtypeWarning.

hissashirocha committed:
Improving docstring for DtypeWarning.

TomAugspurger commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
+
+    See Also
+    --------
+    pd.read_csv : Read CSV (comma-separated) file into a DataFrame.
```
pd -> pandas I think.

hissashirocha replied:
OK!

TomAugspurger commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
+    Important to notice that df2 will contain both `str` and `int` for the
+    same input, '1'.
+
+    >>> df2.iloc[262140,0]
```
PEP8: space after comma

hissashirocha replied:
Done!

TomAugspurger commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
+    ...                    'b':['b']*300000})
+    >>> df.to_csv('test', sep='\t', index=False, na_rep='NA')
+    >>> df2 = pd.read_csv('test', sep='\t')
+    Traceback (most recent call last):
```
I think if yyou add a line like

    >>> df2 = pd.read_csv('test', sep='\t')
    ... # doctest: +SKIP

The doctest might work.

jorisvandenbossche replied:
Yeah, I don't think it will be possible to get Warnings to work with doctest, so adding a `# doctest: +SKIP` is fine

hissashirocha replied:
I'm sorry but I couldn't understand what the doctest: +SKIP is supposed to do. I've added it right after the read_csv and i just got an error. Can you help me?

################################################################################
################################### Doctests ###################################
################################################################################

**********************************************************************
Line 28, in pandas.errors.DtypeWarning
Failed example:
    df2 = pd.read_csv('test', sep=' ')
Expected:
    # doctest: +SKIP
    DtypeWarning: Columns (0) have mixed types...
Got nothing

jorisvandenbossche commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
-    uniform dtypes in a column(s) of a given CSV file.
+    Warning raised when importing different dtypes in a column from a file.
+
+    Raised for a dtype incompatibility. This can happen whenever `pd.read_csv`
```
Can you also use `read_csv` here instead of `pd.read_csv` ? (same below for read_table)

hissashirocha replied:
It's done!

jorisvandenbossche commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
+
+    >>> df = pd.DataFrame({'a':['1']*100000 + ['X']*100000 + ['1']*100000,
+    ...                    'b':['b']*300000})
+    >>> df.to_csv('test', sep='\t', index=False, na_rep='NA')
```
can you add '.csv' for the temp file ? 
I would also leave out the `sep='\t'` and `na_rep='NA'`, as they are not relevant to show illustrate the warning

hissashirocha replied:
Done!

jreback commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
-    Warning that is raised for a dtype incompatibility. This
-    can happen whenever `pd.read_csv` encounters non-
-    uniform dtypes in a column(s) of a given CSV file.
+    Warning raised when importing different dtypes in a column from a file.
```
when reading (not importing)

hissashirocha replied:
Tks!

jreback commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
+    or `pd.read_table` encounter non-uniform dtypes in a column(s) of a given
+    CSV file.
+
+    It only happens when dealing with larger files.
```
instead of this comment, you can add in the Notes section, that this can happen in larger files. Its because the dtype checking happens per chunk that is read.

hissashirocha replied:
Tks!

jreback commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
+    Notes
+    -----
+    Despite the warning, the CSV file is imported with mixed types in a single
+    column. See the examples below to better understand this issue.
```
imported -> read. The dtype of the column will be object.

hissashirocha committed:
Improving docstring for DtypeWarning.

hissashirocha committed:
Merge remote-tracking branch 'upstream/master' into docstring_dtype_warning

Made some adjustments after peer's suggestions.

TomAugspurger commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
+    One way to solve this issue is using the parameter `converters` in the
+    `read_csv` and `read_table` functions to explicit the conversion:
+
+    >>> df2 = pd.read_csv('test', sep='\t', converters={'a': str})
```
FYI, this is still `test` not `test.csv`

jorisvandenbossche replied:
Another thing: I think we should recommend `dtype={'a': str}` instead ?

TomAugspurger reviewed (commented):
@jorisvandenbossche running this doctest will leave behind a `test.csv` file, which isn't ideal.

We can

- convert this to a regular code-block
- Use `io.StringIO`, which is uglier but doesn't leave behind a file
- Add a bit to `os.remove('test.csv')`

Preferences?

jorisvandenbossche commented:
Hmm, yes, that's a general problem though, so let's discuss that somewhere else

jorisvandenbossche commented:
@TomAugspurger see https://github.com/pandas-dev/pandas/issues/20302

jorisvandenbossche commented on pandas/errors/__init__.py:
```diff
@@ -38,9 +38,53 @@ class ParserError(ValueError):
+    ...                    'b':['b']*300000})
+    >>> df.to_csv('test.csv', index=False)
+    >>> df2 = pd.read_csv('test.csv')
+    >>> DtypeWarning: Columns (0) have mixed types... # doctest: +SKIP
```
Can you remove the `>>>` on this line? (it might be you need to move the "# doctest: +SKIP" to the line above after the read_csv)

hissashirocha replied:
@jorisvandenbossche @TomAugspurger I've added "import os" and an "os.remove('test.csv')" at the end of the example, just as you've agreed on #20302 , and also put "# doctest: +SKIP" right before the warning, like this: 

    >>> import os
    >>> df = pd.DataFrame({'a':['1']*100000 + ['X']*100000 + ['1']*100000,
    ...                    'b':['b']*300000})
    >>> df.to_csv('test.csv', index=False)
    >>> df2 = pd.read_csv('test.csv')
    >>> os.remove('test.csv')
    # doctest: +SKIP
    DtypeWarning: Columns (0) have mixed types...

But I got this message on validation:

################################################################################
################################### Doctests ###################################
################################################################################

**********************************************************************
Line 32, in pandas.errors.DtypeWarning
Failed example:
    os.remove('test.csv')
Expected:
    # doctest: +SKIP
    DtypeWarning: Columns (0) have mixed types...
Got nothing


And also, removing ".csv" from to_csv and read_csv, like Tom suggested, raises two possible errors:
>>> os.remove('test.csv'): FileNotFoundError: [Errno 2] No such file or directory: 'test.csv'
>>> os.remove('test'): FileNotFoundError: File b'test' does not exist

What do you guys recommend?

jorisvandenbossche reviewed (commented):
Small comment in addition to the one of @TomAugspurger 

Looks good for the rest!

TomAugspurger committed:
Updated example

- PEP8
- Move warning to "output" section
- Fixed second read example
- Cleanup

jorisvandenbossche committed:
move os import

TomAugspurger commented:
I went back and forth on where to put the import :)

jorisvandenbossche commented:
@TomAugspurger the only problem now is that the actual warning is separated from the code block, which looks a bit strange in the html output .. 
But if don't leave the blank line, you get a failing doctest ..

The alternative you mentioned would be

```
>>> import warnings
>>> with warnings.catch_warnings(record=True) as w:
...     df2 = pd.read_csv('test.csv')
>>> w[0].message
pandas.errors.DtypeWarning('Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.')
```

But I agree this boilerplate is a bit unfortunate and distracts from what we actually want to show

jorisvandenbossche commented:
Let's go ahead with the current solution for now. We can always later look into adding our own doctest-extensions (it should be possible to add doctest options while running our own doctests: eg https://github.com/astropy/pytest-doctestplus/blob/master/pytest_doctestplus/output_checker.py)

jorisvandenbossche commented:
@hissashirocha Thanks for the PR!

TomAugspurger commented:
Ah was just going to push this

```
diff --git a/pandas/errors/__init__.py b/pandas/errors/__init__.py
index 16654f022..ff34df64c 100644
--- a/pandas/errors/__init__.py
+++ b/pandas/errors/__init__.py
@@ -68,8 +68,7 @@ class DtypeWarning(Warning):
     ...                    'b': ['b'] * 300000})
     >>> df.to_csv('test.csv', index=False)
     >>> df2 = pd.read_csv('test.csv')
-
-        DtypeWarning: Columns (0) have mixed types
+    ... # DtypeWarning: Columns (0) have mixed types
```

Looks OK. I'll make a new PR.

jorisvandenbossche commented:
Ah, that's a good one :)

hissashirocha commented:
It was a pleasure! This is my first time contributing to pandas and I really enjoyed it. Thanks for your help!
